src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Patterns.cs

```diff
@@ -142,35 +142,72 @@ private BoundExpression CompareWithConstant(BoundExpression input, BoundExpressi
                 );
         }
 
-        private bool MatchIsIrrefutable(TypeSymbol sourceType, TypeSymbol targetType, bool requiredNullTest)
+        private bool? MatchConstantValue(BoundExpression source, TypeSymbol targetType, bool requiredNullTest)
         {
             // use site diagnostics will already have been reported during binding.
             HashSet<DiagnosticInfo> ignoredDiagnostics = null;
-            switch (_compilation.Conversions.ClassifyBuiltInConversion(sourceType, targetType, ref ignoredDiagnostics).Kind)
-            {
-                case ConversionKind.Boxing:
-                case ConversionKind.ImplicitReference:
-                case ConversionKind.Identity:
-                    return true;
-                default:
-                    return false;
-            }
+            var sourceType = source.Type.IsDynamic() ? _compilation.GetSpecialType(SpecialType.System_Object) : source.Type;
```

Comment:
Yes. There are assertions in callers and in methods called from here (e.g. `ClassifyConversionFromType`)